### PR TITLE
Bismark update proposition

### PR DESCRIPTION
**Pull request by charles-bernard**
Hi Björn, 

Listed below are the adjustments we made to the initial repository:
General Notes:
- Every modules have been tested and thus confirmed as compatible with the last release of Bismark (v0.16.3)
- The mapper tool offers now the possibility to submit built-in genome (in fasta) for the generation of Bismark indexes (if the latters don't already exist)
- The methylation extractor offers now the possibility to generate a genome-wide methylation report for all cytosines in the genome (option –cytosine_report)
- We implemented all the 3 new --ignore options that came with the brand new release of Bismark (--ignore, --ignore_r2, --ignore_3prime, --ignore_3prime_r2)
- We fixed some bugs encountered with the xml and py wrapper of the methylation extractor (due to cheetah template in the command tags of the xml and to the glob functions in the python wrapper)
- We forced the compression of the outputs which are not report files for the methylation extractor
- We rewrote the help sections for better readability for our biologist colleagues
- We also wrapped two other Bismark scripts (deduplicate_bismark and bismark2report) because they are often used in our pipelines. The bismark2report is particularly interesting because it takes as inputs the report files generated by the different Bismark modules and returns an intuitive summary in html format. 

Thank you very much for all your work on Bismark wrapping

### CODE REVIEW

**Comment by bgruening** [comment]
@charles-bernard the other PR from you, whould this be merged before this one here? Or was this an mistake?

**Comment by charles-bernard** [comment]
As you wish Björn, the other one concerns the tool "trim_galore", while this one concerns "bismark". I therefore thought it would be more clear to create two independent PR.

File: tools/bismark/.shed.yml [hunk 1]
```diff
@@ -1,18 +0,0 @@
-categories:
```
Comment (bgruening) [inline tools/bismark/.shed.yml]: We need such a file for every repository in the TS.
This file servers as metadata storage and can be used to upload tools to the TS from the commandline.

File: tools/bismark/bismark_deduplicate/bismark_deduplicate_wrapper.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
+#!/usr/bin/python
```
Comment (bgruening) [inline tools/bismark/bismark_deduplicate/bismark_deduplicate_wrapper.py]: This tool also needs such a .shed.yml file to make the upload automatically.

File: tools/bismark/bismark_deduplicate/bismark_deduplicate_wrapper.xml [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        <exit_code range=":-1" />
+        <regex match="Error:" />
+        <regex match="Exception:" />
+    </stdio>
+
+    <command interpreter="python">
+<![CDATA[
+        bismark_deduplicate_wrapper.py
+
+        --tool_dir $__tool_directory__
```
Comment (bgruening) [inline tools/bismark/bismark_deduplicate/bismark_deduplicate_wrapper.xml]: All path should be quoted like `$__tool_directory__`. How knows maybe someone serves Galaxy in `/my great path/foo/`

File: tools/bismark/bismark_deduplicate/bismark_deduplicate_wrapper.xml [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+        #if str ( $sPaired ) == "single":
+        	-s
+        #else
+        	-p
+        #end if
+
+        --input $mapping_output
```
Comment (bgruening) [inline tools/bismark/bismark_deduplicate/bismark_deduplicate_wrapper.xml]: Quote.

File: tools/bismark/bismark_mapping/bismark_bowtie2_wrapper.xml [hunk 5]
```diff
@@ -0,0 +1,561 @@
...
+    </stdio>
+    <command interpreter="python">
+<![CDATA[
+        bismark_wrapper.py
+
+        ## Change this to accommodate the number of threads you have available.
+        --num-threads "\${GALAXY_SLOTS:-24}"
+
+        ##--bismark_path \$SCRIPT_PATH
+        --bismark_path $__tool_directory__
```
Comment (bgruening) [inline tools/bismark/bismark_mapping/bismark_bowtie2_wrapper.xml]: quote ... and know I think this was my fault ;)

File: tools/bismark/bismark_mapping/bismark_bowtie2_wrapper.xml [hunk 6]
```diff
@@ -0,0 +1,561 @@
+<tool id="bismark_bowtie2" name="Bismark Mapper" version="0.10.2">
```
Comment (bgruening) [inline tools/bismark/bismark_mapping/bismark_bowtie2_wrapper.xml]: Can you please change the tool version to reflect the bismark version, so 0.16.3 here?

File: tools/bismark/bismark_methyl_extractor/bismark_methylation_extractor.xml [hunk 7]
```diff
@@ -0,0 +1,305 @@
+<tool id="bismark_methylation_extractor" name="Bismark Meth. Extractor" version="0.10.2">
```
Comment (bgruening) [inline tools/bismark/bismark_methyl_extractor/bismark_methylation_extractor.xml]: bump of the version number?

File: tools/bismark/bismark_pretty_report/bismark2report_wrapper.xml [hunk 8]
```diff
@@ -0,0 +1,76 @@
+<tool id="bismark_pretty_report" name="Bismark Pretty Report" version="0.10.2">
```
Comment (bgruening) [inline tools/bismark/bismark_pretty_report/bismark2report_wrapper.xml]: version?

File: tools/bismark/tool_dependencies.xml [hunk 9]
```diff
@@ -1,15 +1,38 @@
 <?xml version="1.0"?>
 <tool_dependency>
+	<!-- from test tool shed-->
+    <!--<package name="samtools" version="0.1.19">
+        <repository changeset_revision="a0ab0fae27e5" name="package_samtools_0_1_19" owner="iuc" toolshed="https://testtoolshed.g2.bx.psu.edu" />
+    </package>-->
+    <!--from main tool shed-->
     <package name="samtools" version="0.1.19">
-        <repository name="package_samtools_0_1_19" owner="iuc" />
+        <repository changeset_revision="c9bd782f5342" name="package_samtools_0_1_19" owner="iuc" toolshed="https://toolshed.g2.bx.psu.edu" />
```
Comment (bgruening) [inline tools/bismark/tool_dependencies.xml]: changeset_revision and toolshed should be removed here.

File: tools/bismark/tool_dependencies.xml [hunk 10]
```diff
@@ -1,15 +1,38 @@
...
     </package>
-    <package name="bowtie2" version="2.2.5">
-        <repository name="package_bowtie_2_2_5" owner="iuc" />
-    </package>
-    <package name="bowtie" version="1.0.0">
-        <repository name="package_bowtie_1_0_0" owner="iuc" />
```
Comment (bgruening) [inline tools/bismark/tool_dependencies.xml]:
```
    <package name="bowtie2" version="2.2.5">
         <repository name="package_bowtie2_2_1_0" owner="devteam" />
     </package>
```

File: tools/bismark/bismark_bowtie_wrapper.xml [hunk 11]
```diff
@@ -1,582 +0,0 @@
-<tool id="bismark_bowtie" name="Bismark" version="0.14.3">
```
Comment (bgruening) [inline tools/bismark/bismark_bowtie_wrapper.xml]: Any reason why you removed this tool?

Comment (charles-bernard) [inline tools/bismark/bismark_bowtie_wrapper.xml]: No legit reason to be honest, just that no one in the lab of epigenetics with whom I'm working with would use bowtie 1 for the alignment... Sorry if I deleted it in my development dir and forget to put it back in the PR... but you could re-add it if you like.

**Comment by charles-bernard** [comment]
Thank you so much for all these relevant comments. 
I think I made the appropriate changes...

For the tests I can begin with creating a test-data directory with the dataset I have (Arabidopsis thaliana bisulfite reads) and the resulting Bismark modules outputs.  

Concerning the test section in the xml files, I could give a hand but more likely the next week !

Have a good day !

**Comment by bgruening** [comment]
@charles-bernard thanks for your changes!
For the testing this documentation for planemo will help you: http://planemo.readthedocs.io/en/latest/commands/test.html <- very easy was to run tests for Galaxy tools.

For inspiration of tests see this one here: https://github.com/galaxyproject/tools-devteam/blob/master/tools/bowtie2/bowtie2_wrapper.xml#L584

I'm a little bit worried that the entire Arabidopsis Genome is too big - maybe just a part? Tests data should primarily test the Galaxy tool so the result does not need to be meaningful.

**Comment by bgruening** [comment]
@charles-bernard are you planning to work on the tests? This are awesome changes and I think many users will profit from it.

**Comment by charles-bernard** [comment]
Hi @bgruening ! Sorry for the late answer ! 
These days I'm busy working on a project , but hopefully it should be finished soon... As soon as it is,  I will work on the tests (probably at the beginning of December)

All the best

**Comment by bgruening** [comment]
Thanks this sounds great!

**Comment by bgruening** [comment]
@charles-bernard do you have time to work on this? I guess we should get this out there are so many nice things in it!

**Comment by charles-bernard** [comment]
Dear Bjorn,

I'm truly sorry but my internship on Galaxy at the ENS is over and I would have therefore to work on these tests on my free time, which, I'm afraid, would be difficult as I'm in the middle of a move from Paris to Vienna... But on the other hand, as the current version of bismark has no tests, perhaps you could consider merging the PR as I didn't modify your code but only had new functionalities which worked fine when I tested them.. the latest version of the code is probably this one: https://github.com/charles-bernard/Galaxy_tools/tree/master/bismark.

I'm truly sorry not to be able to contribute more to these tests, especially because it is not the first time you asked me to do so...

All the best,

Charles

**Comment by bgruening** [comment]
@charles-bernard too bad! What are you doing in Vienna and which lab, if any.
Thanks again for this PR!

**Comment by charles-bernard** [comment]
Hi Bjorn ! Hope you are doing well ! I got a scholarship to do an internship at the IST Vienna in Guet's lab.. Does it sound familiar to you?

**Pull request merged**